text/0000-inline-asm.md

```diff
@@ -0,0 +1,869 @@
+| x86 | `reg` | None | `i32` | `eax` |
+| x86 | `reg` | None | `i64` | `rax` |
+| x86-32 | `reg_abcd` | `b` | Any | `al` |
+| x86-64 | `reg` | `b` | Any | `al` |
+| x86 | `reg_abcd` | `h` | Any | `ah` |
+| x86 | `reg` | `w` | Any | `ax` |
+| x86 | `reg` | `k` | Any | `eax` |
+| x86-64 | `reg` | `q` | Any | `rax` |
+| x86 | `vreg` | None | `i32`, `i64`, `f32`, `f64`, `v128` | `xmm0` |
+| x86 (AVX) | `vreg` | None | `v256` | `ymm0` |
+| x86 (AVX-512) | `vreg` | None | `v512` | `zmm0` |
+| x86 (AVX-512) | `kreg` | None | Any | `k1` |
+| AArch64 | `reg` | None | Any | `x0` |
+| AArch64 | `reg` | `w` | Any | `w0` |
+| AArch64 | `reg` | `x` | Any | `x0` |
+| AArch64 | `vreg` | None | Any | `v0` |
+| AArch64 | `vreg` | `b` | Any | `b0` |
+| AArch64 | `vreg` | `h` | Any | `h0` |
+| AArch64 | `vreg` | `s` | Any | `s0` |
+| AArch64 | `vreg` | `d` | Any | `d0` |
+| AArch64 | `vreg` | `q` | Any | `q0` |
+| ARM | `reg` | None | Any | `r0` |
+| ARM | `vreg` | None | `f32` | `s0` |
+| ARM | `vreg` | None | `f64`, `v64` | `d0` |
+| ARM | `vreg` | None | `v128` | `q0` |
+| ARM | `vreg` | `e` / `f` | `v128` | `d0` / `d1` |
+| RISC-V | `reg` | None | Any | `x1` |
+| RISC-V | `vreg` | None | Any | `f0` |
+
+> Notes:
+> - on ARM `e` / `f`: this prints the low or high doubleword register name of a NEON quad (128-bit) register.
+> - on AArch64 `reg`: a warning is emitted if the input type is smaller than 64 bits, suggesting to use the `w` modifier. The warning can be suppressed by explicitly using the `x` modifier.
+
+[llvm-argmod]: http://llvm.org/docs/LangRef.html#asm-template-argument-modifiers
+
+## Flags
+
+Flags are used to further influence the behavior of the inline assembly block.
+Currently the following flags are defined:
+- `pure`: The `asm` block has no side effects, and its outputs depend only on its direct inputs (i.e. the values themselves, not what they point to). This allows the compiler to execute the `asm` block fewer times than specified in the program (e.g. by hoisting it out of a loop) or even eliminate it entirely if the outputs are not used. A warning is emitted if this flag is used on an `asm` with no outputs.
+- `nomem`: The `asm` blocks does not read or write to any memory. This allows the compiler to cache the values of modified global variables in registers across the `asm` block since it knows that they are not read or written to by the `asm`.
+- `readonly`: The `asm` block does not write to any memory. This allows the compiler to cache the values of unmodified global variables in registers across the `asm` block since it knows that they are not written to by the `asm`.
+- `preserves_flags`: The `asm` block does not modify the flags register (defined below). This allows the compiler to avoid recomputing the condition flags after the `asm` block.
+- `noreturn`: The `asm` block never returns, and its return type is defined as `!` (never). Behavior is undefined if execution falls through past the end of the asm code.
```

Comment:
I was actually confused by the name `noreturn`. I initially assumed this meant that there is no `ret` instruction (or similar) in the assembly. How about using `diverging` or `diverge` as the name instead?

Reply:
I feel that `noreturn` expresses this better: execution never returns from the asm block.